LSP 3.18 - Symbol Tags

ratzdi opened the pull request:
Implementation of the issues https://github.com/microsoft/language-server-protocol/pull/2003 and https://github.com/clangd/clangd/issues/2123. Continuation of https://github.com/llvm/llvm-project/pull/113669.

Related Follow-up PR: https://github.com/llvm/llvm-project/pull/170103

chouzz committed:
[clangd] Support symbolTags for document symbol

chouzz committed:
Fix access for class method

chouzz committed:
Support Declaration and Definition tags

chouzz committed:
Remove constant tag

Commit:
Fix test

Commit:
Tagging final methods and classes

Commit:
Checks the extraction of symbol tags

ratzdi committed:
Merge branch 'main' into users/ratzdi/symbol_tags

Commit:
Docu update

Commit:
Fix code format.

ratzdi committed:
Merge branch 'main' into users/ratzdi/symbol_tags

ratzdi committed:
Merge branch 'main' into users/ratzdi/symbol_tags

Commit:
Fill symbol tags for the response CallHierarchyItem of prepareCallHierarchy.

Commit:
Fix call-hierarchy.test

Added tags node in the expected JSON output.

ratzdi committed:
Merge branch 'main' into users/ratzdi/symbol_tags

ratzdi commented:
Hi @kadircet, @llvm-beanz, I would be very happy to see you as a reviewer!

ratzdi committed:
Merge branch 'main' into users/ratzdi/symbol_tags

HighCommander4 commented on clang-tools-extra/clangd/Protocol.h:
```diff
@@ -1104,6 +1104,29 @@ struct CodeAction {
 };
 llvm::json::Value toJSON(const CodeAction &);
 
+enum class SymbolTag {
```
If I'm understanding this correctly, the list and order here is coming from https://github.com/microsoft/language-server-protocol/pull/2003.

Can we link to that PR from a comment here? That makes it clear that we're implementing something that's not in the spec yet, and serves as a reminder for us to update the implementation if the spec ends up being different from the current list in the PR.

ratzdi replied:
Done

HighCommander4 reviewed (changes requested):
Thanks for the PR and sorry for the long turnaround time in responding.

I have two high-level pieces of feedback.

---

First, I think it would be nice to cut down on the amount of new "API" we're exposing in AST.h, as well as some of the code duplication between symbol tags and highlighting modifiers.

(To give an example of what code duplication I'm referring to: it's not obvious why the logic for assigning the `Declaration` tag checks for `UnresolvedUsingValueDecl`. [This comment](https://searchfox.org/llvm/rev/1e985b6ddf023af5782d48c1cce881668fdf6ceb/clang-tools-extra/clangd/SemanticHighlighting.cpp#1177-1179) explains it nicely, but rather than asking for the comment to be duplicated, I'd rather avoid the check itself being duplicated in the first place.)

Here's a specific proposal:

 * Establish a "packed" representation for symbol tags. This can be as simple as `using SymbolTags = uint32_t;` , with the interpretation that if tag `N` is present, then the bit `1 << N` is set in the value.
   * (The reason for having a packed representation at all is so that we don't impose the cost of allocating vectors onto all consumers including semantic highlighting.)
 * Have the public API in `AST.h` just be `SymbolTags computeSymbolTags(Decl*)` or similar.
 * The `FindSymbols` code can call `computeSymbolTags()`, and unpack the result into a `vector<SymbolTag>`.
    * (If we later have multiple requests return a `vector<SymbolTag>`, it's fine to expose a function that returns a `vector<SymbolTag>` in `AST.h` as well.)
 * The `SemanticHighlighting` code can also call `computeSymbolTags()`, and use a mapping from `SymbolTag` to `HighlightingModifier` to convert them to highlighting modifiers. (It's fine to do this only for the subset of modifiers which are also symbol tags. The remaining modifiers can be added individually.)

This way, the logic (including the `UnresolvedUsingValueDecl` check) can remain centralized in the implementation of `computeSymbolTags()`.

How does that sound?

---

Second, regarding testing strategy: given how relatively hard to read lit tests are, we generally use lit tests as a basic "smoke test" that the protocol conversions are working fine, while putting interesting logic into unit tests.

In this case, I think it would be good to have some unit tests that exercise the new `computeSymbolTags()` API. They don't need to be extensive, I think it's sufficient to test on a small piece of code similar to the one that `symbol-tags.test` sends in its `didOpen`, that the computed tags for various symbols are the expected ones.

To make this as easy as possible, I would suggest using the existing `DocumentSymbols` test fixture ([example test](https://searchfox.org/llvm/source/clang-tools-extra/clangd/unittests/FindSymbolsTests.cpp#432)), with a new matcher added around [here](https://searchfox.org/llvm/rev/1e985b6ddf023af5782d48c1cce881668fdf6ceb/clang-tools-extra/clangd/unittests/FindSymbolsTests.cpp#34-37) that checks the symbol tags. (This will exercise the `computeSymbolTags()` API indirectly via its use in the `FindSymbols` feature, but I think that's good enough.) Happy to go into more details about this if it's not clear what I'm suggesting.

ratzdi committed:
Merge branch 'llvm:main' into users/ratzdi/symbol_tags

Commit:
Merge remote-tracking branch 'fork/users/ratzdi/symbol_tags' into users/ratzdi/symbol_tags

Commit:
Reduced AST API, added unit-test introduced symbol-tag bitmask.

Commit:
Use parts of result of the function computeSymbolTags in SemanticHighlighting.

ratzdi committed:
Merge branch 'llvm:main' into users/ratzdi/symbol_tags

HighCommander4 commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -30,6 +30,52 @@
+  const auto SymbolTags = computeSymbolTags(ND);
+  std::vector<SymbolTag> Tags;
+
+  if (SymbolTags & toSymbolTagBitmask(SymbolTag::Deprecated))
```
We should be able to do this with a loop.

HighCommander4 commented on clang-tools-extra/clangd/SemanticHighlighting.h:
```diff
@@ -126,6 +130,47 @@ std::vector<SemanticTokensEdit> diffTokens(llvm::ArrayRef<SemanticToken> Before,
 
+
+// Whether T is const in a loose sense - is a variable with this type readonly?
+bool isConst(QualType T);
```
Part of my motivation in asking not to have these granular functions in `AST.h`, is to avoid having them in any header file. They should be able to live wherever the implementation that computes symbol tags lives, in this case `FindSymbols.cpp`.

HighCommander4 commented on clang-tools-extra/clangd/unittests/FindSymbolsTests.cpp:
```diff
@@ -35,6 +35,8 @@ MATCHER_P(withName, N, "") { return arg.name == N; }
 MATCHER_P(withKind, Kind, "") { return arg.kind == Kind; }
 MATCHER_P(withDetail, Detail, "") { return arg.detail == Detail; }
 MATCHER_P(symRange, Range, "") { return arg.range == Range; }
+MATCHER_P(withSymbolTag, Tag, "") { return llvm::is_contained(arg.tags, Tag); }
```
Instead of doing a separate lookup for each tag (quadratic in the number of tags), we can check them in a single go with:

```c++
Matcher<DocumentSymbol>
withSymbolTags(std::initializer_list<SymbolTag> Tags) {
  return Field(&DocumentSymbol::tags, ElementsAreArray(Tags));
}
```

and then

```c++
withSymbolTags({SymbolTag::Abstract, SymbolTag::Definition})
```

You can also use `UnorderedElementsAreArray` if you'd prefer.

ratzdi replied:
Done

HighCommander4 reviewed (commented):
(I know the patch is a draft but I took a quick look and spotted a few things.)

Commit:
Overload computeSymbolTags to encapsulate the proper determination of Declarations and Definitions.

Commit:
Relocated computeSymbolTags to the module FindSymbols.
Refactoring on getSymbolTags and computeSymbolTags.
Improved SymbolTags test.

Commit:
Fixed tests.

Commit:
Includes clean-up.

Commit:
Removed computeSymbolTags(const NamedDecl &ND) from public API.

HighCommander4 commented on clang-tools-extra/clangd/SemanticHighlighting.h:
```diff
@@ -126,6 +129,10 @@ std::vector<SemanticTokensEdit> diffTokens(llvm::ArrayRef<SemanticToken> Before,
 
+// Indicates whether declaration D is a unique definition (as opposed to a
+// declaration).
+bool isUniqueDefinition(const NamedDecl *Decl);
```
This doesn't need to be exposed in the header.

HighCommander4 commented on clang-tools-extra/clangd/unittests/FindSymbolsTests.cpp:
```diff
@@ -1132,6 +1138,28 @@ TEST(DocumentSymbolsTest, PragmaMarkGroupsNoNesting) {
+TEST(DocumentSymbolsTest, SymbolTags) {
+  TestTU TU;
+  Annotations Main(R"(
+      class A {
```
It would be nice to test at least the amount of code we have in `symbol-tags.test`, i.e.

> class AbstractClass{\n public:\n virtual ~AbstractClass() = default;\n virtual void f1() = 0;\n void f2() const;\n protected: \n void f3(){}\n private: \n static void f4(){} \n }; void AbstractClass::f2() const {} \n class ImplClass final: public AbstractClass { \n public: \n void f1() final {}};"
}}}

If you don't want to have that full code in two places, I'd rather have the shorter code in the lit test and the longer code here.

(Rationale: if we make tweaks to the tag behaviour, we'll want to accompany them with tests, and they'll be more readable as unit tests.)

Commit:
Fix format.

Commit:
Review findings:
Refactored unit and integration tests.
Robust SymbolTag declaration.
Static assert of the symbol tag bit-mask's size.
And other minor improvements...

Commit:
Fix test.

ratzdi commented on clang-tools-extra/clangd/XRefs.cpp:
```diff
@@ -1812,6 +1812,7 @@ declToHierarchyItem(const NamedDecl &ND, llvm::StringRef TUPath) {
   HI.name = printName(Ctx, ND);
   // FIXME: Populate HI.detail the way we do in symbolToHierarchyItem?
   HI.kind = SK;
+  HI.tags = getSymbolTags(ND);
```
This change is intended in #170103. I will delete this line. In this PR, for the first, we collect the symbol tags in the method documentSymbol. More will follow in #170103.

ratzdi commented on clang-tools-extra/clangd/FindSymbols.cpp:
```diff
@@ -23,13 +23,217 @@
+//
+// This is confusing, and maybe we should use another name, but because "static"
+// is a standard LSP modifier, having one with that name has advantages.
+bool isStatic(const Decl *D) {
```
Done.

Commit:
Review findings:
refactored computeSymbolTags - no IsDecl is needed,
refactored getSemanticHighlightings - intro map for tags and modifiers,

HighCommander4 commented on clang-tools-extra/clangd/Protocol.h:
```diff
@@ -1513,6 +1548,9 @@ struct TypeHierarchyItem {
   SymbolKind kind;
 
+  /// The symbol tags for this item.
+  std::vector<SymbolTag> tags;
```
Per previous discussion, this change to `TypeHierarchyItem` shouldn't be in this patch

HighCommander4 reviewed (changes requested):
Thanks -- looking pretty good, just minor comments remaining.

Commit:
Review findings:
- obsolete includes,
- public toSymbolTagBitmask,
- removed tags from TypeHierarchyItem,
- thread_local map in getSemanticHighlightings
- minor improvements

Commit:
Fix format.

HighCommander4 reviewed (approved):
Thanks! LGTM with final nits (drop changes to AST.h and SemanticHighlighting.h).

Commit:
Review findings:
- removed obsolete forward-decl and includes
- restored include

ratzdi committed:
Merge branch 'main' into users/ratzdi/symbol_tags

HighCommander4 commented on clang-tools-extra/clangd/SemanticHighlighting.h:
```diff
@@ -22,7 +22,6 @@
 
 #include "Protocol.h"
 #include "llvm/ADT/StringRef.h"
-#include "llvm/Support/raw_ostream.h"
```
There's still a change to `SemanticHighlighting.h` here.

(I would have fixed this for you myself and merged, but it looks like "Maintainers are allowed to edit this pull request" is not checked, so I can't 🤷 )

Commit:
Restored include.
